### PR TITLE
Fix 5% regression in techempower plaintext after change to SocketOutput

### PR DESCRIPTION
**Pull request by halter73**
The main improvement is that the _contextLock is never reacquired on the threadpool to return written blocks. This seems to make a significant difference on better hardware.

#1145 

The improvement on a 48 logical core machine is pretty significant. Unfortunately, I'm getting RPS numbers as high as those @nathana1 mentioned in the issue before or after the change. I wonder if there's been another regression since @nathana1, or if I'm doing something wrong measuring.

@nathana1 I left my session open on the server if you want to check it out. Thanks.

### dev

    $ ./multi-run.sh --depth 16
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    4470347.84
    Running tests...
    4453736.01
    4453718.90
    4420181.20
    4434269.48
    4508274.92
    4488297.48
    4399945.67
    4466734.33
    4593534.31
    4531329.83
    4453736.01 4453718.90 4420181.20 4434269.48 4508274.92 4488297.48 4399945.67 4466734.33 4593534.31 4531329.83

### halter73/1145

    $ ./multi-run.sh --depth 16
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    4699696.24
    Running tests...
    4758701.48
    4733957.38
    4707075.01
    4707825.09
    4753937.18
    4763750.12
    4673806.49
    4691102.53
    4703620.58
    4753383.33
    4758701.48 4733957.38 4707075.01 4707825.09 4753937.18 4763750.12 4673806.49 4691102.53 4703620.58 4753383.33

### CODE REVIEW

**Comment by halter73** [comment]
@nathana1 I reran the tests before and after the SocketOutput regression and compared the % reduction/improvement vs this PR. This seems to me that this PR makes up for 75% of the original 2.5% regression or about 0.6% total RPS. It's not shocking that the extra synchronization required to return all blocks when idle has a small perf impact.

What's also notable about these number is that there seems to have been another commit since the once the introduced the SocketOutput regression that's made the RPS numbers even worse.

#### 0edf36b (Prior to SocketOutput regression)

    $ ./multi-run.sh --depth 16 5069879 RPS
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    5047311.80
    Running tests...
    5136810.64
    5085791.75
    5070092.00
    5068440.00
    5066827.52
    5043093.36
    5079766.22
    5079208.79
    5061944.23
    5006810.58
    5136810.64 5085791.75 5070092.00 5068440.00 5066827.52 5043093.36 5079766.22 5079208.79 5061944.23 5006810.58

#### 7b2f7b9 (SocketOutput regression commit) 4944431 RPS (~2.5% reduction)

    $ ./multi-run.sh --depth 16
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    4878541.79
    Running tests...
    4943812.54
    4937836.98
    4946996.03
    4917154.86
    4958126.67
    4937056.60
    4935910.18
    4908845.84
    4981247.88
    4977320.41
    4943812.54 4937836.98 4946996.03 4917154.86 4958126.67 4937056.60 4935910.18 4908845.84 4981247.88 4977320.41

#### 9149762 (dev) 4864990 Avg RPS

    $ ./multi-run.sh --depth 16
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    4796551.14
    Running tests...
    4874085.57
    4858118.69
    4872137.35
    4882243.08
    4836012.78
    4863452.72
    4877888.01
    4880109.75
    4852677.87
    4853172.72
    4874085.57 4858118.69 4872137.35 4882243.08 4836012.78 4863452.72 4877888.01 4880109.75 4852677.87 4853172.72

#### 8b88cfa halter73/1145 4957118 Avg RPS (~1.9% improvement)

    $ ./multi-run.sh --depth 16
    ./run-wrk.sh -p 5000 -h 192.168.1.100 -t plaintext --script ../pipeline.lua -d 1m --depth 16 -c 512
    Warming up...
    4939704.99
    Running tests...
    4988249.40
    4922895.88
    4929945.02
    5004944.76
    4965914.35
    4978231.66
    4925067.84
    4933531.63
    4952750.78
    4969648.92
    4988249.40 4922895.88 4929945.02 5004944.76 4965914.35 4978231.66 4925067.84 4933531.63 4952750.78 4969648.92

**Comment by nathana1** [comment]
I didn't see any new regressions since you commit through https://github.com/aspnet/KestrelHttpServer/commit/12adc7de9dfe6e325626dc930429749bd6964527.  Do any of the commits following that look suspicious to you?

**Comment by halter73** [comment]
No particular commit is too concerning. But neither was the SocketOutput change, but 9149762 since that is updating Kestrel's dependencies to Microsoft.NETCore.App 1.1.0-*. This has since been reverted due to AVs that seem to be a crossgen bug, but all my most recent numbers are with Microsoft.NETCore.App 1.1.0-*. The other changes between 12adc7d and 9149762 are mostly validation related. There's a small risk it hurt perf, but I find it unlikely.

@nathana1 If you compare  halter73/1145 (8b88cfa) to the previous commit (9149762), do you not see an improvement?

**Comment by nathana1** [comment]
halter73/1145 is a definite improvement, about 2.5% over current dev branch.  Any second regression looks much smaller (maybe 1%) perhaps https://github.com/aspnet/KestrelHttpServer/commit/4117ad8a1ee9b0ea532884fd1f9e0c437f246569  but I don't see any reason why in that commit.

**Comment by benaadams** [comment]
`Interlocked.Read` was introduced in  https://github.com/aspnet/KestrelHttpServer/commit/0312da7df39874e7f3554ec556d3203f1385e343 replaced with with Volatile.Read in https://github.com/aspnet/KestrelHttpServer/pull/1153 might be unrelated though.

**Comment by halter73** [comment]
@nathana1 Thanks.

**Comment by halter73** [comment]
@cesarbs Want to give this a review?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 1]
```diff
@@ -694,40 +697,55 @@ private void PoolWriteReq(UvWriteReq writeReq)
...
-                    block = block.Next;
-
-                    returnBlock.Pool.Return(returnBlock);
-                }
+                var returnAll = false;
 
                 lock (Self._returnLock)
                 {
-                    // If everything has been fully written, return _tail.
+                    // If everything has been fully written, return _tail/_lockeEnd.
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: uber nit: `_lockeEnd` -> `_lockedEnd`

Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: You got eagle eyes! At this point you know to expect at least one spelling error in my PRs. 😝

**Comment by halter73** [comment]
ce6e65b

**Pull request closed without merging**